### PR TITLE
when edge register into cloud, cloud send all info to edge

### LINKED ISSUE #260: If edge core didn't handle pod delete event from cloud, pod remains with status Terminating
**What happened**:

Pod (running at edge) remain with status Terminating.

image

For temporary go over this problem, I have to delete meta entry of the pod in edge.db and delete the pod object in etcd manully.

**What you expected to happen**:

Pod should be deleted.

**How to reproduce it (as minimally and precisely as possible)**:

Deploy pod to edge, after pod up and running, kill edge core, delete the pod, again start the edge core.

Then the pod we intend to delete get never deleted, ramains with status Terminating.

**Anything else we need to know?**:

The downstream controller (response for trigger event to edge side) just try to notice edge once, and if failed did nothing but log an error.

https://github.com/kubeedge/kubeedge/blob/6878d86b9ec33ca716933575b93ef70663504ba5/cloud/edgecontroller/pkg/controller/controller/downstream.go#L51-L67

**Environment**:
- KubeEdge version: v0.2
- Hardware configuration: irrelevant
- OS (e.g. from /etc/os-release): irrelevant
- Kernel (e.g. `uname -a`): irrelevant
- Others: irrelevant

### PR DESCRIPTION
**Pull request by kadisi**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind test
> /kind failing-test
> /kind feature

/kind bug

**What this PR does / why we need it**:

when edgecontroller first started , and  edge_core sencond started after many hours or days.
edgecontroller first get all pods and then send to the relevant host channelPool
```
type ChannelEventQueue struct {
	ctx         *context.Context
	channelPool sync.Map
}
```
but if now , edge_core not started, the relevant host channelpool can not find, so edgecontroller can not send full ammount event to edge_core, this will result in loss of pod information.

so we need save these events in advance.

but maybe we need merge this events to Remove old duplicates.
**Which issue(s) this PR fixes**:

Fixes #455 
FIxes #260

**Special notes for your reviewer**:

### CODE REVIEW

File: cloud/pkg/cloudhub/channelq/channelq.go [hunk 1]
```diff
@@ -110,15 +115,14 @@ func (q *ChannelEventQueue) getRChannel(nodeID string) (chan model.Event, error)
...
-	}
-	// allocate a new rchannel with default buffer size
-	rChannel := make(chan model.Event, rChanBufSize)
-	_, ok = q.channelPool.LoadOrStore(info.NodeID, rChannel)
-	if ok {
-		// rchannel is already allocated
-		return fmt.Errorf("edge node %s is already connected", info.NodeID)
+	if !ok {
+		// allocate a new rchannel with default buffer size
+		rChannel := make(chan model.Event, rChanBufSize)
```
Comment (sids-b) [inline cloud/pkg/cloudhub/channelq/channelq.go]: There is a problem here, if ok is true, we will return nil which is wrong. ok means edgenode with same id already exists and it should return an error. Please undo this change

File: cloud/pkg/cloudhub/channelq/channelq.go [hunk 2]
```diff
@@ -84,6 +84,11 @@ func (q *ChannelEventQueue) dispatchMessage() {
 			log.LOGGER.Warnf("node id is not found in the message")
 			continue
 		}
+		_, ok := q.channelPool.Load(nodeID)
+		if !ok {
+			rChannel := make(chan model.Event, rChanBufSize)
```
Comment (sids-b) [inline cloud/pkg/cloudhub/channelq/channelq.go]: We should not create channel of node which are not connected.

**Comment by CindyXing** [comment]
@kadisi , thanks for raising the issue. Can you clarify the customer impact here? Are we having a data loss? Can the edgenode get pod information later when channelpool created?

**Comment by kadisi** [comment]
@CindyXing , in cloud, controller will create DownstreamController which create many manager, such as PodManager, ConfigManager. In the case of PodManager,  it will get realEvents by SharedInformer and ResourceEventHander 
```
// NewPodManager create PodManager from config
func NewPodManager(kubeClient *kubernetes.Clientset, namespace string) (*PodManager, error) {
	lw := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "pods", namespace, fields.Everything())
	realEvents := make(chan watch.Event)
	mergedEvents := make(chan watch.Event)
	rh := NewCommonResourceEventHandler(realEvents)
	si := cache.NewSharedInformer(lw, &v1.Pod{}, 0)
	si.AddEventHandler(rh)

	pm := &PodManager{realEvents: realEvents, mergedEvents: mergedEvents}

	stopNever := make(chan struct{})
	go si.Run(stopNever)
	go pm.merge()

	return pm, nil
}
```
when edgecontroller start, it will get all pod events(first list all pods, and then watch them) , and merge them into `mergedEvents channel`.

and in `syncPod` function, it will Consuming these events。and send them by messageLayer.Send
```
func (dc *DownstreamController) syncPod(stop chan struct{}) {
...
			if err := dc.messageLayer.Send(*msg); 
...
}

```

after send , cloudhub module will receive this msg by Context.
```
func (q *ChannelEventQueue) dispatchMessage() {
	for {
		msg, err := q.ctx.Receive("cloudhub")
		if err != nil {
			log.LOGGER.Infof("receive not Message format message")
			continue
		}
		resource := msg.Router.Resource
		tokens := strings.Split(resource, "/")
		numOfTokens := len(tokens)
		var nodeID string
		for i, token := range tokens {
			if token == "node" && i+1 < numOfTokens {
				nodeID = tokens[i+1]
				break
			}
		}
		if nodeID == "" {
			log.LOGGER.Warnf("node id is not found in the message")
			continue
		}
		rChannel, err := q.getRChannel(nodeID)
		if err != nil {
			log.LOGGER.Infof("fail to get dispatch channel for %s", nodeID)
			continue
		}
		event := model.MessageToEvent(&msg)
		select {
		case rChannel <- event:
		default:
		}
	}
}
```

but in this function,  it need get host channel , but  it can not get that host channel, because edge_core not started.
if we continue, and do nothing , we will lost that whole amount of information. only get the watched information after edge_core started.

**Comment by sids-b** [comment]
@kadisi , till edge_core starts, the node will be in notReady state and hence scheduler will not schedule any event for that node. So I think there will not be any events for that node. Thoughts ?

**Comment by kadisi** [comment]
@sids-b yes , scheduler will not schedule any event for that node.  but if we keep the old logic,，edge_core only get the watched information after  edge_core starts.    edge_core can not get enough pod information , and can not clean up the dirty pod

**Comment by sids-b** [comment]
@kadisi , to get cleanup pod information, we need something like reliable message delivery in cloud. Also one more reason I am suggesting to not make a channel if node is not connected is we may not know for how long a node is disconnected, so the channel buffer may get full and we may end up blocked ? Thoughts ?

**Comment by CindyXing** [comment]
> @CindyXing , in cloud, controller will create DownstreamController which create many manager, such as PodManager, ConfigManager. In the case of PodManager, it will get realEvents by SharedInformer and ResourceEventHander
> 
> ```
> // NewPodManager create PodManager from config
> func NewPodManager(kubeClient *kubernetes.Clientset, namespace string) (*PodManager, error) {
> 	lw := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "pods", namespace, fields.Everything())
> 	realEvents := make(chan watch.Event)
> 	mergedEvents := make(chan watch.Event)
> 	rh := NewCommonResourceEventHandler(realEvents)
> 	si := cache.NewSharedInformer(lw, &v1.Pod{}, 0)
> 	si.AddEventHandler(rh)
> 
> 	pm := &PodManager{realEvents: realEvents, mergedEvents: mergedEvents}
> 
> 	stopNever := make(chan struct{})
> 	go si.Run(stopNever)
> 	go pm.merge()
> 
> 	return pm, nil
> }
> ```
> 
> when edgecontroller start, it will get all pod events(first list all pods, and then watch them) , and merge them into `mergedEvents channel`.
> 
> and in `syncPod` function, it will Consuming these events。and send them by messageLayer.Send
> 
> ```
> func (dc *DownstreamController) syncPod(stop chan struct{}) {
> ...
> 			if err := dc.messageLayer.Send(*msg); 
> ...
> }
> ```
> 
> after send , cloudhub module will receive this msg by Context.
> 
> ```
> func (q *ChannelEventQueue) dispatchMessage() {
> 	for {
> 		msg, err := q.ctx.Receive("cloudhub")
> 		if err != nil {
> 			log.LOGGER.Infof("receive not Message format message")
> 			continue
> 		}
> 		resource := msg.Router.Resource
> 		tokens := strings.Split(resource, "/")
> 		numOfTokens := len(tokens)
> 		var nodeID string
> 		for i, token := range tokens {
> 			if token == "node" && i+1 < numOfTokens {
> 				nodeID = tokens[i+1]
> 				break
> 			}
> 		}
> 		if nodeID == "" {
> 			log.LOGGER.Warnf("node id is not found in the message")
> 			continue
> 		}
> 		rChannel, err := q.getRChannel(nodeID)
> 		if err != nil {
> 			log.LOGGER.Infof("fail to get dispatch channel for %s", nodeID)
> 			continue
> 		}
> 		event := model.MessageToEvent(&msg)
> 		select {
> 		case rChannel <- event:
> 		default:
> 		}
> 	}
> }
> ```
> 
> but in this function, it need get host channel , but it can not get that host channel, because edge_core not started.
> if we continue, and do nothing , we will lost that whole amount of information. only get the watched information after edge_core started.

Thanks @kadisi for the details. This is a great scenario and can be a common one. My thought is that after edgecore is started, send something to edgecontroller and be able to get the delta; afterwards is the watch. It can be timestamp of last successful sync. 

In general, we want to make sure edge and cloud in sync and control the message load between.

**Comment by CindyXing** [comment]
> @kadisi , to get cleanup pod information, we need something like reliable message delivery in cloud. Also one more reason I am suggesting to not make a channel only for cleanup is if node is not connected is we may not know for how long a node is disconnected, so the channel buffer may get full and we may end up blocked ? Thoughts ?

Agree with @sids-b that the problem is how we can ensure the reliable message delivery between cloud and edge; and be in sync. Let's address the big one than target specifics.

**Comment by kadisi** [comment]
> This is a great scenario and can be a common one. My thought is that after edgecore is started, send something to edgecontroller and be able to get the delta; afterwards is the watch. It can be timestamp of last successful sync.

@sids-b @CindyXing  you are right,   as @CindyXing  said, when edgecore is started, she need tell edgecontroller that she has started and she need get the detail information about that host.

**Comment by CindyXing** [comment]
> > This is a great scenario and can be a common one. My thought is that after edgecore is started, send something to edgecontroller and be able to get the delta; afterwards is the watch. It can be timestamp of last successful sync.
> 
> @sids-b @CindyXing you are right, as @CindyXing said, when edgecore is started, she need tell edgecontroller that she has started and she need get the detail information about that host.

@kadisi The scenario is valid. I think we all agree. It is just the current approach needs to be discussed. Do you attend the bi-weekly meeting? We can chat there or through github.

**Comment by kadisi** [comment]
@CindyXing  is bi-weekly meeting on Wednesday at 11:00 ? i think i am ok

**Comment by kadisi** [comment]
@CindyXing @sids-b i think we need to cache these changing events on cloudhub side.
in cloudhub, we store change events in `channelPool` 
```
// ChannelEventQueue is the channel implementation of EventQueue
type ChannelEventQueue struct {
	ctx         *context.Context
	channelPool sync.Map
}
```

but `channelPool` is map, it`s value is no-buffer-channel,   we need to change this channel to a queue to catch this changeing event.

 and this queue need support upstate event function 。 if we receive a new event which has the same attributes(Group, Source,UserGroup, ID, ParenetID) as an event int he queue. we only update the event in the queue.  If the new event is different from all the event properties in the queue, we need push it to the queue.
